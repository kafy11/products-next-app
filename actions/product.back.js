import generateProductDAO from '../daos/productDAO';

//product actions exclusive for the back-end

export const getProducts = async () => {
    const { getAllProducts } = generateProductDAO();
    return await getAllProducts();
};

export const getProduct = async (product_id) => {
    const { getProduct } = generateProductDAO();
    return await getProduct(product_id);
};