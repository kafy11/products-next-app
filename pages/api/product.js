import generateProductDAO from '../../daos/productDAO';

const { insertProduct, getAllProducts } = generateProductDAO();

export default async (req, res) => {
    switch(req.method){
        case 'POST':
            await insertProduct(req.body);
            res.status(200).json({ success: 1 });
            break;
        
        case 'GET':
            const products = await getAllProducts();
            res.status(200).json(products);
            break;
        
        default:
            res.status(405).setHeader('Allow', 'POST, GET').send('Not Allowed');
    }
}
