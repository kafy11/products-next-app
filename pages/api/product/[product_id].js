import generateProductDAO from '../../../daos/productDAO';

const { updateProduct, getProduct, deleteProduct } = generateProductDAO();

export default async (req, res) => {
    const {
        query: { product_id },
    } = req;

    switch(req.method){
        case 'POST':
            await updateProduct({
                product_id,
                ...req.body
            });
            res.status(200).json({ success: 1 });
            break;
        
        case 'GET':
            try{
                const product = await getProduct(product_id);
                res.status(200).json(product);
            } catch (error){
                //if query doesn't return a product
                res.status(404).json('Product not found')
            }
            break;
        
        case 'DELETE':
            await deleteProduct(product_id);
            res.status(200).json({ success: 1 });
            break;
        
        default:
            //blocks all other methods
            res.status(405).setHeader('Allow', 'POST, GET, DELETE').send('Not Allowed');
    }
}
