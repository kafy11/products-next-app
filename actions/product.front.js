import { callPostAPI, callDeleteAPI } from '../libs/api';

//product actions for the pages

export const addProduct = async (product) => {
    await callPostAPI('product', product);
};

export const updateProduct = async (product_id, product) => {
    await callPostAPI(`product/${product_id}`, product);
};

export const deleteProduct = async (product_id) => {
    await callDeleteAPI(`product/${product_id}`);
};