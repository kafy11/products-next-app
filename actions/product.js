import { callPostAPI, callGetAPI, callDeleteAPI } from '../libs/api';

//product actions for the pages

export const addProduct = async (product) => {
    await callPostAPI('product', product);
};

export const updateProduct = async (product_id, product) => {
    await callPostAPI(`product/${product_id}`, product);
};

export const deleteProduct = async (product_id) => {
    await callDeleteAPI(`product/${product_id}`);
};

export const getProducts = async () => {
    const products = await callGetAPI('product');
    return products;
};

export const getProduct = async (product_id) => {
    const product = await callGetAPI(`product/${product_id}`);
    return product;
};