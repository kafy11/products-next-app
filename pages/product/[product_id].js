import ProductFormPage from '../../templates/ProductFormPage';
import { useRouter } from 'next/router';
import { updateProduct, getProducts, getProduct } from '../../actions/product';

const EditProductPage = ({ product_id, ...rest }) => {
    const router = useRouter();

    const handleSubmit = async (product) => {
        await updateProduct(product_id, product);
        router.push('/');
    }

    return (
        <ProductFormPage 
            title="Create Product"
            onSubmit={handleSubmit}
            {...rest}
        />
    );
}

//get the products to generate all possible paths
export async function getStaticPaths() {
    const products = await getProducts(); 
    const paths = products.map(({ product_id }) => `/product/${product_id}`);
  
    return { paths, fallback: false };
}

//get the product to pre-render the page
export async function getStaticProps({ params }) {
    const product = await getProduct(params.product_id);
  
    return {
        props: {
            ...product,
        },
    }
  }

export default EditProductPage;