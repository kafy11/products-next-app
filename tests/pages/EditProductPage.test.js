import { shallow } from 'enzyme';
import EditProductPage from '../../pages/product/[product_id]';
import products from '../fixtures/products';
import { updateProduct } from '../../actions/product';

const mockPush = jest.fn();
jest.mock('next/router', () => ({
    useRouter: () => ({
        push: mockPush
    })
}));

jest.mock('../../actions/product', () => ({
    updateProduct: jest.fn(),
}));

const product = products[0];

let wrapper;
beforeEach(() => {
    wrapper = shallow(<EditProductPage {...product} />);
});

describe('EditProductPage', () => {
    it('should render correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });
    
    it('should pass the product info to the ProductFormPage', () => {
        const formWrapper = wrapper.find('ProductFormPage');
    
        expect(formWrapper.prop('name')).toBe(product.name);
        expect(formWrapper.prop('price')).toBe(product.price);
        expect(formWrapper.prop('quantity')).toBe(product.quantity);
    });
    
    it('should update the product and redirect to the list when submitting form', async () => {
        const { product_id, ...rest } = product;
        await wrapper.find('ProductFormPage').prop('onSubmit')(rest);
    
        expect(updateProduct).toHaveBeenLastCalledWith(product_id, rest);
        expect(mockPush).toHaveBeenLastCalledWith('/');
    });
});