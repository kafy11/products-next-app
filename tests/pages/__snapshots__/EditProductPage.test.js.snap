// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditProductPage Component should render correctly 1`] = `
<ProductFormPage
  name="PS4"
  onSubmit={[Function]}
  price={399}
  quantity={1}
  title="Create Product"
/>
`;
